### DISCUSSION TITLE
Caching of static arguments

### DISCUSSION THREAD

**Discussion by cisprague** [jax-ml/jax#20998] [General]
I was searching the internet/documentation to see if `jax.jit` retains a cache for previously used static arguments to a function. I couldn't seem to find any information, so I did a quick test. It seems that `jax.jit` does retain the cache.

```python
import jax, jax.numpy as jnp
from functools import partial
x = jax.random.normal(jax.random.PRNGKey(0))
```

```python
class Function:
    @staticmethod
    @partial(jax.jit, static_argnums=(1,))
    def __call__(x, y):
        jnp.exp(x) if y else jnp.exp(x)
f = Function()
```

```python
%timeit -n 100 -r 100 f(x, True)
%timeit -n 100 -r 100 f(x, False)
%timeit -n 100 -r 100 f(x, True)
%timeit -n 100 -r 100 f(x, False)
```

```
The slowest run took 8.54 times longer than the fastest. This could mean that an intermediate result is being cached.
12.3 µs ± 6.15 µs per loop (mean ± std. dev. of 100 runs, 100 loops each)
The slowest run took 9.37 times longer than the fastest. This could mean that an intermediate result is being cached.
14.1 µs ± 7.08 µs per loop (mean ± std. dev. of 100 runs, 100 loops each)
13.7 µs ± 3.89 µs per loop (mean ± std. dev. of 100 runs, 100 loops each)
13.4 µs ± 3.95 µs per loop (mean ± std. dev. of 100 runs, 100 loops each)
```

**Comment by jakevdp** [comment]
Yes, `jit` has a runtime cache so that subsequent calls to a JIT-compiled function will hit the cache rather than recompile. The cache key is constructed from the shape and dtype of input arrays, the hash of static arguments, and global configuration values that may affect execution. The details are internal, but there are a few APIs that let you inspect the cache; for example:
```python
import jax

def f(x, flag):
  return x if flag else x + 1

f_jit = jax.jit(f, static_argnums=1)

print(f_jit._cache_size())  # 0

f_jit(1.0, True)
print(f_jit._cache_size())  # 1

f_jit(1.0, False)
print(f_jit._cache_size())  # 2

# re-wrapped function hits the same cache
f_jit_2 = jax.jit(f, static_argnums=1)
print(f_jit_2._cache_size())  # 2

# cache hits don't increase the cache size
f_jit(100.0, True)
f_jit(100.0, False)
print(f_jit._cache_size())  # 2
```

**Reply by cisprague** [reply]
Thanks for the info @jakevdp! How do things work when `vmap` gets involved? 

The use case I have in mind is that I would like to define a class that has methods with a nice API, but the methods use jitted static methods under the hood. It seems like this is achievable with:

```python
import jax, jax.numpy as jnp
from functools import partial

x0 = jax.random.normal(jax.random.PRNGKey(0), (50,))
x1 = jax.random.normal(jax.random.PRNGKey(1), (50, 2))
```

```python
class Function:

    def __call__(self, x):
        return self.g(x)
        
    @staticmethod
    @jax.jit
    def g(x):
        if x.ndim == 1:
            return Function._g(x)
        else:
            return jax.vmap(Function._g)(x)
        
    @staticmethod
    @jax.jit
    def _g(x):
        # assume that this function is too complicated
        # to define in Function.g
        return jnp.sum(x)
```

```python
f = Function()
print(f.g._cache_size()) # 0
f(x0)
print(f.g._cache_size()) # 1
f(x1)
print(f.g._cache_size()) # 2
f(x0)
print(f.g._cache_size()) # 2
f(x1)
print(f.g._cache_size()) # 2
```

But, I would prefer to do is this:

```python
class Function:

    def __call__(self, x):
        if x.ndim == 1:
            return self.g(x)
        else:
            # Don't want to make method for this
            return jax.vmap(self.g)(x)
        
    @staticmethod
    @jax.jit
    def g(x):
        return jnp.sum(x)
```

But this gives `0, 1, 1, 1, 1` for the above execution. Any ideas on how to improve this?

**Reply by jakevdp** [reply]
`vmap` of `jit` will not hit any cache, but `jit` of `vmap` will hit a cache. So if you want your vmapped function to hit a JIT cache, make sure `jit` is on the outside.

**Reply by cisprague** [reply]
Right, but if we do the following, it seems to hang for quite a long time.

```python
class Function:

    def __call__(self, x):
        if x.ndim == 1:
            return jax.jit(self.g)(x)
        else:
            return jax.jit(jax.vmap(self.g))(x)
        
    @staticmethod
    def g(x):
        return jnp.sum(x)
```

**Reply by cisprague** [reply]
@jakevdp, apologies if this seems unrelated, but I believe it is in the same vein as what is discussed here. I have now tried to go the PyTree node route:

```python
@register_pytree_node_class
class Model:

    def __init__(self, square: bool, other: bool) -> None:
        self.square = square # Used in __call__
        self.other = other # Not used in __call__

    @jax.jit
    def __call__(self, x: float) -> float:
        return x**2 if self.square else x

    def tree_flatten(self):
        return (tuple(), (self.square, self.other))

    @classmethod
    def tree_unflatten(cls, static, dynamic):
        return cls(*dynamic, *static)
```
My intention here is that `__call__` should only recompile if a relevant static argument changes, but this is not the case:

```python
m = Model(True, True)
print(m.__call__._cache_size()) # 0
m(2.0)
print(m.__call__._cache_size()) # 1
m(2.0)
print(m.__call__._cache_size()) # 1

m.other = False
m(2.0)
print(m.__call__._cache_size()) # 2 (should be 1)
m(2.0)
print(m.__call__._cache_size()) # 2 (should be 1)
```
Any idea how to ameliorate this? In my code, I am using child classes as well, so keeping things in a class would be nice.
Thanks!

**Reply by jakevdp** [reply]
> Right, but if we do the following, it seems to hang for quite a long time.

The reason for that is that you're calling `jit` on a dynamically-created function `vmap(f)` within the function call, so you're never hitting the JIT cache, and are therefore paying the recompilation cost on every call. The fix would be to JIT-compile the outer `__call__` function rather than calling `jit` within its implementation.

> My intention here is that __call__ should only recompile if a relevant static argument changes, but this is not the case:

This is working as expected. Whether or not you're using a pytree, the JIT cache is based on all the static arguments to a function, even if the function does not use some or all of those arguments (the reason for this is that the only way to know whether a function's behavior changes based on argument values is to run the function!) In your case, you have two implicit static arguments (the two static values returned in tree flattening) and if either of those changes the function will be re-compiled. If this is not the behavior you want, then you could call out to another JIT-compiled function that only takes as arguments the values it uses.

**Reply by cisprague** [reply]
Thanks for the response. I would like recompilation only to happen when an attribute relevant to the method changes. And I would like to do this without having to define anything outside of the class. In my case, I want to define a JITed parent class method that uses a child method, where the child method itself dynamically and statically makes use of some combination of its own attributes and its parent's attributes. Shouldn't this be possible by just comparing the method signature to the set of statically-treated attributes?

**Reply by jakevdp** [reply]
> Shouldn't this be possible by just comparing the method signature to the set of statically-treated attributes?

I don't think this is possible. Say you have a function that looks like this:
```python
def f(x, condition1, condition2):
  return x + 1 if condition1 else x + 2
```
How would you propose determining that the value of `condition2` does not affect the output of the function, without executing it to see whether the results change?

**Reply by cisprague** [reply]
@jakevdp thanks for your help. It seems I'm able to get the behaviour I want in the following way.

```python
@dataclass
class Parent:

    dp: float # dynamic parameter
    sp: bool # static parameter

    def kwargs(self):
        return (dict(dp=self.dp), dict(sp=self.sp))

    def __call__(self, x: float) -> float:
        # API method
        dynamic, static = self.kwargs()
        f = jax.jit(self.f, static_argnames=('g', *static.keys()))
        return f(x, self.g, **(dynamic|static))
    
    @staticmethod
    def f(x: float, g: Callable, **kwargs) -> float:
        # Jitted parent method
        return kwargs['dp']*g(x, **kwargs) if kwargs['sp'] else g(x, **kwargs)
    
    @staticmethod
    def g(x):
        # To be implemented in child classes
        raise NotImplementedError
    
@dataclass
class Child(Parent):

    dc: float # dynamic child parameter
    sc: bool # static child parameter

    def kwargs(self):
        dynamic, static = super().kwargs()
        return (dynamic|dict(dc=self.dc), static|dict(sc=self.sc))

    @staticmethod
    def g(x, **kwargs):
        return kwargs['dc']*x if kwargs['sc'] else x
```

```python
c = Child(2.0, False, 3.0, False)
%time c(1.0) # Compile
%time c(1.0) # No compile
print('\nChanging static args...')
c.sp = True  # Static arg change
%time c(1.0) # Compile
%time c(1.0) # No compile2
print('\nChanging static args...')
c.sc = True  # Static arg change
%time c(1.0) # Compile
%time c(1.0) # No compile
print('.\nChanging dynamic args...')
c.dp = 3.0   # Dynamic arg change
%time c(1.0) # No compile!
%time c(1.0) # No compile!
print('\nChanging dynamic args...')
c.dc = 4.0   # Dynamic arg change
%time c(1.0) # No compile!
%time c(1.0) # No compile!
```
I don't really like using a dictionary to handle this, but it seems to be a solution. What do you think? Does this already exist?

**Reply by jakevdp** [reply]
Yeah, that seems reasonable, though I'd probably prefer to wrap `g` in `jit` directly where it's defined

**Reply by cisprague** [reply]
Yeah, I would like to do the JITing where the method is defined, but it seems that the partitioning of the dynamic and static arguments is not accessible outside of the method definitions. Maybe we could use the data class field metadata as a workaround. It seems like something similar has been done in the dataclass-type decorators of Flax, Chex, and TJax. But I'm not sure if they would allow changes to attributes in the way I want.
